ui/components/app/signature-request/signature-request.component.js

```diff
@@ -45,9 +45,85 @@ export default class SignatureRequest extends PureComponent {
       hardwareWalletRequiresConnection,
     } = this.props;
     const { address: fromAddress } = fromAccount;
-    const { message, domain = {} } = JSON.parse(data);
+    const { message, domain = {}, types, primaryType } = JSON.parse(data);
     const { metricsEvent } = this.context;
 
+    const mapType = (msgType, definedType) => {
+      // try and map it to a solidity type
+      if (
+        (definedType.type.indexOf('int') >= 0 ||
+          definedType.type.indexOf('fixed') >= 0) &&
+        msgType === 'number'
+      ) {
+        return true;
+      } else if (
+        (definedType.type.indexOf('bytes') >= 0 ||
+          definedType.type === 'address') &&
+        msgType === 'string'
+      ) {
+        return true;
+      }
+
+      return false;
+    };
+
+    const sanitizeMessage = (msg, baseType) => {
+      if (version === 'V4') {
+        const sanitizedMessage = {};
+        const msgKeys = Object.keys(msg);
+        const msgTypes = Object.values(msg).map((value) => typeof value);
+        const baseTypeType = types[baseType];
+        msgKeys.forEach((msgKey, index) => {
+          const valueType = msgTypes[index];
+          if (baseTypeType) {
```

Comment:
This doesn't need to be in the loop. I'd suggest checking this before the loop, and negating this condition and exiting early rather than indenting everything.